test/gpu/native/studies/sort/radixSortGpu.chpl

```diff
@@ -15,101 +16,198 @@ module RadixSort {
+      arr[hostArr.domain] = hostArr;
+      var offset = 1;
+      while offset < arr.size {
+        var arrBuffer = arr;
+        @assertOnGpu
+        foreach idx in 0..<arr.size/(2*offset) {
+          const doubleOff = offset << 1;
+          const i = idx*doubleOff;
+          arr[i+doubleOff-1] = arrBuffer[i+offset-1] + arrBuffer[i+doubleOff-1];
         }
+        offset = offset << 1;
+      }
 
-        // Down-sweep
-        prefixSum[buckets-1] = 0;
-        offset = buckets >> 1;
-        while offset > 0 {
-          var prefixSumBuffer = prefixSum;
-          @assertOnGpu
-          foreach idx in 0..<buckets/(2*offset) {
-            const i = idx*2*offset;
-            const t = prefixSumBuffer[i+offset-1];
-            prefixSum[i+offset-1] = prefixSumBuffer[i+2*offset-1];
-            prefixSum[i+2*offset-1] = prefixSum[i+2*offset-1] + t;
-          }
-          offset = offset >> 1;
+      // Down-sweep
+      arr[arr.size-1] = 0;
+      offset = arr.size >> 1;
+      while offset > 0 {
+        var arrBuffer = arr;
+        @assertOnGpu
+        foreach idx in 0..<arr.size/(2*offset) {
+          const i = idx*2*offset;
+          const t = arrBuffer[i+offset-1];
+          arr[i+offset-1] = arrBuffer[i+2*offset-1];
+          arr[i+2*offset-1] = arr[i+2*offset-1] + t;
         }
-        hostprefixSum = prefixSum;
+        offset = offset >> 1;
+      }
+      hostArr=arr[hostArr.domain];
+    }
+  }
+
+  proc arrScan(ref arr: [] uint){
+    if scanType == "hillisSteele" then
+      hillisSteeleScan(arr);
+    else if scanType == "blelloch" then
+      blellochScan(arr);
+    else if scanType == "serial" then
+      serialScan(arr);
+  }
+
+  proc parallelArrScan(ref arr: [] uint){
+    // Divide up the array into chunks of a reasonable size
+    // Config this with a scanChunkSize
+    const numScanChunks = Math.divCeil(arr.size,scanChunkSize);
+
+    if numScanChunks == 1 {
+      arrScan(arr);
+      return;
+    }
+
+    // Allocate an accumulate array
+    var scanArr : [0..<numScanChunks] uint;
+
+    var host = here;
+
+    on here.gpus[useGpuId]{
+      var gpuScanArr = scanArr;
+      var gpuArr = arr;
+      // In parallel: For each chunk we do an in lane serial scan
+      // @assertOnGpu
+      for chunk in 0..<numScanChunks {
+        const start = chunk*scanChunkSize;
+        const end = min(gpuArr.domain.high, start+scanChunkSize-1);
+        gpuScanArr[chunk] = gpuArr[end]; // Save the last element before the scan overwrites it
+        serialScan(gpuArr, start, end); // Exclusive scan in serial
+        gpuScanArr[chunk] += gpuArr[end]; // Save inclusive scan in the scan Arr
+
       }
+
+      on host {
+        // Scan the scanArr
+        // This is risky because gpuScanArr lives on the GPU
+        // 2 out of 3 scan algorithms also work on GPU
+        // So it's fine for those
+        // But the serialScan is on the CPU
+        // It will crash and burn if we run that one
```

Comment:
Could we do some sort of assertion at runtime instead of crash-and-burn?

Reply:
Sure!